src/ccache.c

```diff
@@ -2634,6 +2634,11 @@ cc_process_args(struct args *args, struct args **preprocessor_args,
 
 		// Debugging is handled specially, so that we know if we can strip line
 		// number info.
+		if (str_eq(argv[i], "-g0")) {
```

Comment:
This is not quite enough since `-g3 -g0` still will enter the if statement below and for instance set `conf->run_second_cpp = true`.

And more importantly: since `-g0` isn't added to `stripped_args` it won't be sent to the compiler so the compiler will then use `-g3`.

Reply:
Yes, there's also another problem with:
1. conf->unify = true
2. -g1 was parsed and conf->unify was set to false
3. -g0 was parsed, but we can't restore conf->unify back to true, as there is no information of state of conf->unify before the -g1 caused it to be set to false

How do we solve this?

Reply:
Right. Thinking more about this, I now remember that we used to have smarter processing of `-g*` options (#92) but it was reverted (b8aeb9826fcbf4627df8c0d9a9b110c4523ea0f1) due to being buggy (#149).

The best solution I see is to remember the last `-g*` option and only act (maybe disable unify, maybe set , etc.) after we have looped over all arguments. But it is quite complex to figure out how all options interact, so it's not clear to me how that should be implemented. For instance, `-ggdb3 -g0` seems to be equivalent to `-g3 -ggdb0`, but `-ggdb3 -g2` is not equivalent to `-g3 -ggdb2`...

So maybe we should keep things safe and not try to handle `-g3 -g0` like I suggested. We could still handle the simple case in #368 like you did, but the argument then needs to be added to `stripped_args`.

Hmm... It does seem like `-g*0` disables all debug info enabled by previous options, though, so maybe we could make use of that fact. I made an attempt in 749fa11f3d59 on a separate branch. What do you think about that approach?

Reply:
Looks good to me. Will you push it?

Reply:
Thanks for having a look. I'll push it after writing some tests.